### ISSUE TITLE
In-Page Inserter Revisions

### ISSUE THREAD

**Issue by mtias** [WordPress/gutenberg#10519]
There are a few refinements that had been discussed in the last few weeks around the state of in-page inserters (the trailing inserter, the empty block inserter, the sibling inserter, the bottom of the page click area, etc).

These have all evolved through multiple feedback loops and they are in a pretty good state. However, there are a few refinements that keep coming back that we should look at doing or punting within the next week or so.

### Goals

To reiterate, the main goals for these insertion mechanisms are:

- Consistent interface for adding blocks.
- Provide shortcuts for common interactions.
- - Adding blocks between other blocks.
- - Adding paragraphs and media.

The combination of these goals is why we have a few inconsistencies like the sibling inserter using a `+` icon yet not opening the inserter but adding an empty paragraph instead. Or why we have a trailing inserter with shortcuts for commonly accessed blocks.

### Tasks

The proposed iterations, as I'm gathering them, would be:

- **Sibling Inserter**: Either open the full inserter or change the icon to be a paragraph, it doesn't help to have a `+` button that behaves differently to the others as it causes uncertainty.
- **Trailing Inserter**: Always show a `+` at the bottom of the page. This would be similar to the new block appender for non-text contexts: https://github.com/WordPress/gutenberg/pull/10136
- **Bottom of the Page**: Is this something we want to keep if there is always a visible `+` at the end?

Combining quick access to some blocks (paragraph and image) could be done like this too:

![always-there-plus-centered](https://user-images.githubusercontent.com/548849/46812903-ff20af80-cd75-11e8-9abf-bf8723dd8b1c.gif)

**Comment by StaggerLeee** [comment]
There is no option to delete inserted block with "Add text or type...", unless you make it real block and than delete.
It is not thoughtful done.

Imagine User clicks on it by mistake. He/She does not want any block, just clicked and activated it by mistake. Then forcing User to choose some real block just to be able to delete it is wrong. At least by my opinion.

You had it nicely done once and changed for some strange reason. When User activate inserter, and clicked outside this block Inserter DIV (or call it pre-block) is automatically and silently deleted.

Make it work at least with Del keyboard key, when dis DIV is empty anyway.

**Comment by chrisvanpatten** [comment]
Hi @StaggerLeee, that very issue is being tracked over in #10334. I'd recommend checking out that ticket.

**Comment by StaggerLeee** [comment]
Thank you @chrisvanpatten. User @kkoppenhaver said it fine and I agree, as it is shown im my edit.

**Comment by aduth** [comment]
I've opened #11329 to start exploring considerations for how the always-visible trailing appender should behave.

**Comment by afercia** [comment]
Noticed this issue just because it was mentioned in https://github.com/WordPress/gutenberg/issues/7177#issuecomment-429398736. Please add the "Accessibility" label to any issue or PR that touches UI controls or other parts of the UI that impact accessibility. 

Re: the specific proposal illustrated in the GIF above, I'd like to know how this design addresses accessibility of the "appearing" controls for keyboard users, screen reader users, speech input users, users with low vision (contrast ratio), users with cognitive impairments, etc. Thank you.

**Comment by StaggerLeee** [comment]
Please bring back inserter at the end of the content. Why did you remove it ?

**Comment by mtias** [comment]
@afercia yes, it'd have to be a through re-evaluation of "insertion at the end". The gif just shows a more clear area of "end of post / insert more content" that is not tied to the paragraph. I would expect it to be of similar affordances for keyboard users, screen reader users, etc, as a single place for "add more content", with "recent blocks" being potentially an addition to it, if you want to, but with the focus on the `+` action.

**Comment by afercia** [comment]
Reading https://github.com/WordPress/gutenberg/pull/12510 and https://github.com/WordPress/gutenberg/pull/12536#issuecomment-443662139 I'd be in favor of entirely removing the current "inserter shortcuts" as there are a few a11y issues. 

First, the visual order doesn't match the DOM order so when tabbing with the keyboard, the tab sequence "jumps" from the right to the left.

Right now, there's also a contrast issue. These shortcuts are styled differently in two different contexts. In the "default appender", some CSS rules override the default styling of the `InserterWithShortcuts` component and the icons do have a sufficient contrast. The used color is `$dark-opacity-500` (`$light-opacity-500` for dark themes):

<img width="996" alt="default appender" src="https://user-images.githubusercontent.com/1682452/50054965-c2dc4780-0148-11e9-8236-8c230c8d1ec6.png">

When clicking, the default appender becomes the "default block" and the `InserterWithShortcuts` component uses its default styling. In this case, the contrast is not sufficient. The used color is now $dark-opacity-light-700` ($light-opacity-light-700` for dark themes):

<img width="1001" alt="default block" src="https://user-images.githubusercontent.com/1682452/50055010-4ac25180-0149-11e9-86d1-5465596524bc.png">

Also, after https://github.com/WordPress/gutenberg/pull/12510 the default appender shortcuts are rendered only on hover, so [the CSS opacity transition](https://github.com/WordPress/gutenberg/blob/a617aeefb8359a5a12ca20ca3ef00681d1f814e1/packages/editor/src/components/default-block-appender/style.scss#L29) doesn't work any longer. Just noting it because right now there's some CSS that doesn't do anything and should be cleaned up when removing these shortcuts. /Cc @jasmussen 

In the meantime, I'd propose to temporarily fix the contrast issue. Will push a quick PR.

**Comment by jasmussen** [comment]
Thanks for the ping Andrea. Given the scope limitation to component directories, I would think we can refactor that whole thing and remove any old CSS once we redo this thing. I'd be reluctant to over optimize for what we have today, given they might be going soon.

**Comment by kjellr** [comment]
_Swapped out `Needs Design Feedback` for `Needs Design` on this one — since it looks like we're at a stage where design explorations would be helpful._

**Comment by aduth** [comment]
In an effort to try to revive this conversation, I've opened #19045 which proposes to remove the current block inserter "shortcut" options.

**Comment by ZebulanStanphill** [comment]
Should this issue still be open? The in-page inserter shortcuts have all been removed now, and the in-page inserter now puts the "+" icon on the right regardless of whether it is as the top level or in a nested block.

**Comment by aduth** [comment]
@ZebulanStanphill As I understand, while #19045 addressed part of the concern here, there are still considerations for additional changes:

- Should the inserter always be shown at the end of the block list?
   - Currently, it is only shown if the last block is not the default (paragraph) block
- Should the clickable area below the block list be removed or revised?

**Comment by ZebulanStanphill** [comment]
Personally, I think the inserter should never be shown at the end of the block list, or at least it should always be shown. Either way would be better than it continuing to be inconsistent.

Personally, I think the clickable area below the block list is kind of confusing, though it doesn't bother me that much.

**Comment by jasmussen** [comment]
> Personally, I think the inserter should never be shown at the end of the block list, or at least it should always be shown. Either way would be better than it continuing to be inconsistent.

I would very much echo something needs to be done here, because the way it is now breaks with the "deselected content is a preview" a fair bit. Here's some some exploratory reusable block work I've been doing:

![Screenshot 2020-02-03 at 18 39 48](https://user-images.githubusercontent.com/1204802/73676453-aedd0780-46b4-11ea-96b8-c135eb7e8b0e.png)

The plus is not shown after a text block, only after a non text block. So the three pluses are a) the one after a separator, b) the one after a spacer, and c) the permanent bottom appender.

If this was _consistent_ (both following text and other blocks), and invisible on a deselected block, it might be fine. But as it is, these buttons are incredibly disruptive to any layout that isn't just text.

**Comment by mapk** [comment]
It appears the display of inserters have undergone significant revisions with the G2 UI. I'd like to close this as these inserters are no longer littered throughout the interface.

I tested using a Group block, a Columns block, a Buttons block, etc. I don't see inserters all over the place anymore which is a HUGE improvement. 🎉 

![Screen Shot 2020-03-16 at 3 11 49 PM](https://user-images.githubusercontent.com/617986/76804022-7b53c980-6798-11ea-8031-b65272213894.png)

The only time I see a lingering inserter is on the reusable block once a block is converted to one.

![Screen Shot 2020-03-16 at 3 14 03 PM](https://user-images.githubusercontent.com/617986/76804165-d7b6e900-6798-11ea-8e75-04e644ec3f4e.png)

For these reasons, and alongside the recent Block Library work being done, I'm going to close this out.
